### PR TITLE
Igorduda 8 count char occurence

### PR DESCRIPTION
**Pull request by IgorDuda**
Please review my issue № 8

### CODE REVIEW

File: php/Tests/StringsTest.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+ */
+class StringsTest extends TestCase
+{
+	
+	/**
+	 * This method is called before a test is executed.
+	 * 
+	 * @return void
+	 */
+	protected function setUp()
```
Comment (p5ych0) [inline php/Tests/StringsTest.php]: not required

File: php/Tests/StringsTest.php [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	{
+		
+	}
+	
+	/**
+	 * This method is called after a test is executed.
+	 * 
+	 * @return void
+	 */
+	protected function tearDown()
```
Comment (p5ych0) [inline php/Tests/StringsTest.php]: same

File: php/Tests/StringsTest.php [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	 * 
+	 * @param type $expected
+	 * @param type $haystack
+	 * @param type $needle
+	 * @dataProvider charOccurrenceProviderSuccess
+	 * @covers ::countCharOccurrence
+	 */
+	public function testSuccessCountCharOccurrence($expected, $haystack, $needle)
+	{
+		$object = new Utils\Strings($haystack);
```
Comment (p5ych0) [inline php/Tests/StringsTest.php]: use 4 spaces instead of tabs

File: php/Tests/StringsTest.php [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	 * 
+	 * @param type $haystack
+	 * @param type $needle
+	 * @dataProvider charOccurrenceProviderException
+	 * @covers ::countCharOccurrence
+	 */
+	public function testExceptionCountCharOccurrence($haystack, $needle)
+	{
+		$this->expectException(\InvalidArgumentException::class);
+		$object = new Utils\Strings($haystack);
```
Comment (p5ych0) [inline php/Tests/StringsTest.php]: answer the question, could you assign a protected property if for example constructor doesn't accept strings?

File: php/Utils/Strings.php [hunk 5]
```diff
@@ -8,6 +8,15 @@ class Strings
      * @var string $string
      */
     protected $string;
+	
+	/**
+	 * 
+	 * @param string $string
+	 */
+	function __construct(string $string)
```
Comment (p5ych0) [inline php/Utils/Strings.php]: imagine there's another constructor which accepts an object and nobody knows what is assigned inside its body

File: php/Utils/Strings.php [hunk 6]
```diff
@@ -22,15 +31,19 @@ public function kebabCase(): string
...
-     * Count char occurences in the given string
+     * Count char occurrences in the given string
      *
      * @param  string $char Needle
-     * @return int Count of occurence
+     * @return int Count of occurrence
      * @throws \InvalidArgumentException
      */
-    public function countCharOccurence(string $char): int
+    public function countCharOccurrence(string $char): int
```
Comment (p5ych0) [inline php/Utils/Strings.php]: nice catch

File: php/Utils/Strings.php [hunk 7]
```diff
@@ -22,15 +31,19 @@ public function kebabCase(): string
...
     {
-        // @TODO: implement body and cover with tests
+		if(is_string($char) && strlen($char) == 1)  {
+			return substr_count($this->string, $char);
```
Comment (p5ych0) [inline php/Utils/Strings.php]: will this support UTF-8?

Comment (IgorDuda) [reply php/Utils/Strings.php]: at this moment not, it should support it or throw exeption in this case ?

File: php/Utils/Strings.php (hunk 6 above)
Comment (IgorDuda) [reply php/Utils/Strings.php]: :+1:

File: php/Utils/Strings.php (hunk 5 above)
Comment (IgorDuda) [reply php/Utils/Strings.php]: Added this validation

File: php/Tests/StringsTest.php (hunk 1 above)
Comment (IgorDuda) [reply php/Tests/StringsTest.php]: removed

File: php/Tests/StringsTest.php (hunk 2 above)
Comment (IgorDuda) [reply php/Tests/StringsTest.php]: removed

File: php/Tests/StringsTest.php (hunk 3 above)
Comment (IgorDuda) [reply php/Tests/StringsTest.php]: Done

File: php/Utils/Strings.php [hunk 8]
```diff
@@ -22,22 +32,34 @@ public function kebabCase(): string
...
-     * @param  string $char Needle
-     * @return int Count of occurence
      * @throws \InvalidArgumentException
      */
-    public function countCharOccurence(string $char): int
+    public function countCharOccurrence(string $char): int
     {
-        // @TODO: implement body and cover with tests
+		if (!is_string($this->string)) {
+            throw new \InvalidArgumentException('Property should be type string');
```
Comment (p5ych0) [inline php/Utils/Strings.php]: basically it's a DomainException, but let's assume `$string` is always a string, however the length could be different

File: php/Utils/Strings.php [hunk 9]
```diff
@@ -22,22 +32,34 @@ public function kebabCase(): string
...
+     * @param string $char Needle
+     *
+     * @return int Count of occurrence
      *
-     * @param  string $char Needle
-     * @return int Count of occurence
      * @throws \InvalidArgumentException
      */
-    public function countCharOccurence(string $char): int
+    public function countCharOccurrence(string $char): int
```
Comment (p5ych0) [inline php/Utils/Strings.php]: we're living in the modern world so have to support multibyte alphabets

File: php/Utils/Strings.php [hunk 10]
```diff
@@ -22,22 +32,34 @@ public function kebabCase(): string
...
     {
-        // @TODO: implement body and cover with tests
+		if (!is_string($this->string)) {
+            throw new \InvalidArgumentException('Property should be type string');
+        }
+		
+        if (is_string($char) && strlen($char) == 1) {
```
Comment (p5ych0) [inline php/Utils/Strings.php]: first check is not required since TypeError introduced and thrown

File: php/Utils/Strings.php [hunk 11]
```diff
@@ -5,15 +5,25 @@
...
-     * @var string $string
+     * @var string
      */
     protected $string;
 
     /**
-     * Change case of a given string to kebeb
+     * @param string $string
+     */
+    public function __construct(string $string)
```
Comment (p5ych0) [inline php/Utils/Strings.php]: don't implement the constructor please

File: php/Utils/Strings.php [hunk 12]
```diff
@@ -22,22 +32,34 @@ public function kebabCase(): string
...
-    public function countCharOccurence(string $char): int
+    public function countCharOccurrence(string $char): int
     {
-        // @TODO: implement body and cover with tests
+		if (!is_string($this->string)) {
+            throw new \InvalidArgumentException('Property should be type string');
+        }
+		
+        if (is_string($char) && strlen($char) == 1) {
+            return substr_count($this->string, $char);
```
Comment (p5ych0) [inline php/Utils/Strings.php]: this must be case-insensitive (sorry)

File: php/Tests/StringsTest.php [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+    /**
+     * @param int    $expected
+     * @param string $string
+     * @param string $char
+     * @dataProvider charOccurrenceProviderSuccess
+     * @covers ::countCharOccurrence
+     */
+    public function testSuccessCountCharOccurrence(int $expected, string $string, string $char)
+    {
+        $object = new Strings();
```
Comment (p5ych0) [inline php/Tests/StringsTest.php]: have heard about mock and reflections?

File: php/Utils/Strings.php [hunk 14]
```diff
@@ -5,15 +5,43 @@
...
+    }
+
+    /**
+     * Perform lower case folding on a string.
+     *
+     * @param string $string
+     *
+     * @return string
+     */
+    private function stringToLowercase(string $string): string
```
Comment (p5ych0) [inline php/Utils/Strings.php]: makes no sense

File: php/Utils/Strings.php [hunk 15]
```diff
@@ -30,15 +58,23 @@ public function kebabCase(): string
      */
     public function countCharOccurrence(string $char): int
     {
-        // @TODO: implement body and cover with tests
-        // Note: upper- and lowercase letters are considered the same character,
+        if (mb_strlen($char) == 1) {
+            $char = $this->stringToLowercase($char);
+            $string = $this->stringToLowercase($this->string);
```
Comment (p5ych0) [inline php/Utils/Strings.php]: no need to (re)assign those vars

File: php/Utils/Strings.php [hunk 16]
```diff
@@ -30,15 +58,23 @@ public function kebabCase(): string
      */
     public function countCharOccurrence(string $char): int
     {
-        // @TODO: implement body and cover with tests
-        // Note: upper- and lowercase letters are considered the same character,
+        if (mb_strlen($char) == 1) {
```
Comment (p5ych0) [inline php/Utils/Strings.php]: logically drop function in the beginning, e.g. `if (mb_strlen($char) !== 1) { throw`

File: php/Utils/Strings.php [hunk 17]
```diff
@@ -5,15 +5,43 @@
...
+    /**
+     * Perform lower case folding on a string.
+     *
+     * @param string $string
+     *
+     * @return string
+     */
+    private function stringToLowercase(string $string): string
+    {
+        return mb_convert_case($string, MB_CASE_LOWER);
```
Comment (p5ych0) [inline php/Utils/Strings.php]: also there's another useful function

Comment (p5ych0) [reply php/Utils/Strings.php]: there's already `mb_strtolower()` so the function doesn't make a sense

**Pull request closed without merging**